Add inventory recommendation on how to loop over hosts (#34)
redhat-cop/automation-good-practices: inventories/README.adoc

            └── locations.yml
----
====

== Rely on your inventory to loop over hosts, don't create lists of hosts
[%collapsible]
====
Explanations::

> When there is the need to do the same task on multiple hosts, don't create a variable with a list of hosts and loop over it.

Reviewer: ```suggestion
To perform the same task on multiple hosts, don't create a variable with a list of hosts and loop over it.
```

Instead use as much as possible the capabilities of your inventory, which is already a kind of list of hosts.
+

> The anti-pattern is especially seen when it comes to the provisioning of hosts against some kind of manager (e.g. a list of VMs against an hypervisor manager, like oVirt/RHV or vCenter, or against a management tool like Foreman/Satellite, or our beloved AWX/Tower/controller).

Reviewer: ```suggestion
The anti-pattern is especially obvious in the example of provisioning hosts on some kind of manager.
Commonly seen automation tasks of this kind are spinning up a list of VMs via a hypervisor manager like oVirt/RHV or vCenter, or calling a management tool like Foreman/Satellite or even our beloved AWX/Tower/controller.
```

Rationale::
There are 4 main reasons for following this advice:
+

> . a list of hosts is more difficult to maintain than an inventory structure, and tends to become very quickly difficult to oversee.

Reviewer: ```suggestion
. a list of hosts is more difficult to maintain than an inventory structure, and it tends to become very difficult to oversee very quickly .
```

Author: ```suggestion
. a list of hosts is more difficult to maintain than an inventory structure, and tends it quickly to become difficult to oversee.
```

> It is especially true as you generally need to maintain your hosts in your inventory.

Reviewer: ```suggestion
This is especially true as you generally already need to maintain your hosts in your inventory.
```

This brings us to the 2nd advantage:
. you avoid duplicating information, as you often need the same kind of information in your inventory that you also need in order to provision your VMs.
In your inventory, you can also use groups to define group variables, automatically inherited by hosts.

> You can try to implement a similar inheritance pattern with your list of hosts, but it becomes difficult and hand-crafted.

Reviewer: ```suggestion
You can try to implement a similar inheritance pattern with your list of hosts, but it quickly becomes difficult and _hand-crafted_.
```

> . as you loop through the hosts of an inventory, Ansible helps you with https://docs.ansible.com/ansible/latest/user_guide/playbooks_strategies.html[parallelization, throttling, etc], all kind of things you can't do with your own list (well, you can combine https://docs.ansible.com/ansible/latest/user_guide/playbooks_async.html[async and loop] to reach something like this, but it's a lot more complex to handle).

Reviewer: ```suggestion
. as you loop through the hosts of an inventory, Ansible helps you with https://docs.ansible.com/ansible/latest/user_guide/playbooks_strategies.html[parallelization, throttling, etc], all of which you can't do easily with your own list (technically, you _can_ combine https://docs.ansible.com/ansible/latest/user_guide/playbooks_async.html[async and loop] to reach something like this, but it's a lot more complex to handle than letting Ansible do the heavy lifting for you).
```

[source,yaml]
----
include::inventory_loop_hosts/playbook_bad.yml[]
----
+
TIP: check the resulting files using e.g. `head -n-0 /tmp/bad_*`.
+
As said, no way to limit the hosts provisioned, and no parallelism.
Compare then with the recommended approach, with a slightly different structure:
+
.Content of the "good" `groups_and_hosts` file
[source,ini]
----
include::inventory_loop_hosts/inventory_good/groups_and_hosts[]
----
+
It is now the hosts and their groups which carry the relevant information, it is not anymore parked in one single list (and can be used for other purposes):
+
.The "good" variable structure
----
$ cat inventory_good/host_vars/host1/provision.yml 
provision_value: uno
$ cat inventory_good/group_vars/managed_hosts_a/provision.yml 
manager_hostname: manager_a
----
+
And the provisioning playbook now runs in parallel and can be limited to specific hosts:
+
.The "good" way to loop over hosts
[source,yaml]
----
include::inventory_loop_hosts/playbook_good.yml[]
----
+
The result isn't overwhelming in this simple setup but you would of course better appreciate if the provisioning would take half an hour instead of a fraction of seconds:
+

> .Comparaison of the execution times between good and bad

Reviewer: ```suggestion
.Comparison of the execution times between the "good" and the "bad" implementation
```

Author: Aah, my French shining through...

----
$ ANSIBLE_STDOUT_CALLBACK=profile_tasks \
	ansible-playbook -i inventory_bad playbook_bad.yml 
Saturday 23 October 2021  13:11:45 +0200 (0:00:00.040)       0:00:00.040 ****** 
Saturday 23 October 2021  13:11:45 +0200 (0:00:00.858)       0:00:00.899 ****** 
=============================================================================== 
create some file to simulate an API call to provision a host ------------ 0.86s
$ ANSIBLE_STDOUT_CALLBACK=profile_tasks \
	ansible-playbook -i inventory_good playbook_good.yml 
Saturday 23 October 2021  13:11:55 +0200 (0:00:00.040)       0:00:00.040 ****** 
Saturday 23 October 2021  13:11:56 +0200 (0:00:00.569)       0:00:00.610 ****** 
=============================================================================== 
create some file to simulate an API call to provision a host ------------ 0.57s
----
+

> TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`. Not so bad...

Reviewer: ```suggestion
TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`.
```

Author: That was supposed to be a funny reference to the 3rd code example. Not so funny?
